js/style/style_layer/fill_style_layer.js

```diff
@@ -6,11 +6,23 @@ const FillBucket = require('../../data/bucket/fill_bucket');
 class FillStyleLayer extends StyleLayer {
 
     getPaintValue(name, globalProperties, featureProperties) {
-        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
-            return super.getPaintValue('fill-color', globalProperties, featureProperties);
-        } else {
-            return super.getPaintValue(name, globalProperties, featureProperties);
+        if (name === 'fill-outline-color') {
+            const target = this.getPaintProperty('fill-outline-color');
+            if (target === undefined) {
+                return super.getPaintValue('fill-color', globalProperties, featureProperties);
+            }
+
+            // https://github.com/mapbox/mapbox-gl-js/issues/3657
+            const transition = this._paintTransitions['fill-outline-color'];
+            if (!transition.instant()) {
+                const source = transition.oldTransition._calculateTargetValue(globalProperties, featureProperties);
```

Comment:
This is an unfortunate use of private method `_calculateTargetValue`, but my feeling is that it's (slightly) less bad than the only other non-invasive alternative that I could see, which would be to hack [interpolate.color](https://github.com/mapbox/mapbox-gl-js/blob/master/js/util/interpolate.js#L27-L32) so that it can handle undefined values for `from`.  I prefer the hack here because it's more localized, and will be easier to excise when we get rid of fill-outline-color.